### ISSUE TITLE
'ForceAuthn' => true is not propagated to the SAMLRequest

### ISSUE THREAD

**Issue by tauceti2** [simplesamlphp/simplesamlphp#641]
Hi all,

When I set 'ForceAuthn' => true in authsources.php under the 'default-sp' then it doesn't affect the SAMLRequest which doesn't contain the forceAuth flag. Is there something else needs to be configured? I'm using 1.14 version.

**Comment by jaimeperez** [comment]
Hi @tauceti2!

I have just tried with the latest stable, and I see it appear fine in the SAML request:

```XML
<samlp:AuthnRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
                    xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
                    ID="_7dd5e8618cf4bf94995835269b811d8d3101644c9a"
                    Version="2.0"
                    IssueInstant="2017-06-22T10:29:25Z"
                    Destination="***"
                    ForceAuthn="true"
```

Could it be a typo in the configuration? Or maybe something else? What exact version are you using?

**Comment by tauceti2** [comment]
Here is the snippet from the conf:

'default-sp' => array(
                'saml:SP',

                # Doesn't work!!!!
                'ForceAuthn' => true,

                // The entity ID of this SP.
                // Can be NULL/unset, in which case an entity ID is generated based on the metadata URL.
                'entityID' => 'https://login.ceitec.cz/proxy/',

And here is the AuthnRequest:

<samlp:AuthnRequest xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
                    xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
                    ID="_e41eff1c7d9b69b96c46c77646f435a740aea8700d"
                    Version="2.0"
                    IssueInstant="2017-06-22T11:22:04Z"
                    Destination="https://login.ceitec.cz/google/saml2/idp/SSOService.php"
                    AssertionConsumerServiceURL="https://login.ceitec.cz/proxy/module.php/saml/sp/saml2-acs.php/default-sp"
                    ProtocolBinding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
                    >
    <saml:Issuer>https://login.ceitec.cz/proxy/</saml:Issuer>
    <samlp:NameIDPolicy Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient"
                        AllowCreate="true"
                        />
    <samlp:Scoping>
        <samlp:RequesterID>http://adfs.ceitec.cz/adfs/services/trust</samlp:RequesterID>
    </samlp:Scoping>
</samlp:AuthnRequest>

**Comment by jaimeperez** [comment]
Are you using the latest 1.14.14 version?

I can't reproduce this, so it must be some local issue. Maybe a typo somewhere?

**Comment by tauceti2** [comment]
I used the code from the github master branch (cca 2 months ago). I don't know if it can have any influence, but I'm using it in proxy SP/IdP mode.

**Comment by jaimeperez** [comment]
Oh, yeah, that definitely makes a difference. I have checked and indeed in proxy mode it doesn't honor that config option, it will just relay what the original SP requested. So if you want to force authentication, it must be the original SP requesting that, otherwise it won't enforce it itself.

I'm reopening the issue, but marking it as a feature request, as I don't think this is really a bug.

**Comment by thijskh** [comment]
Does it even make sense to have this in proxy config? If the original SP does not send it, already SSP may reuse its session to that SP, not even sending a request to the IdP at all...

Can you specify more explicitly what you want to achieve?

**Comment by tauceti2** [comment]
I'd like to solve a problem, when user log out from the SP which log out user also from the Proxy IdP/SP, but not from the IdP which was used to authenticate against the Proxy IdP/SP. Because Proxy IdP/SP has its own session management there is no need to have SSO also towards the user's IdP and forceAuthn make sense here.

**Comment by jaimeperez** [comment]
No, I'd say it doesn't. `ForceAuthn` is intended to enforce authentication, not to resolve misbehaviour during logout. When you log out of the SP, a logout request is sent to the proxy, which should end its session and propagate its logout to the IdP. If the IdP is not logging out, there's an issue in the IdP. If it is the SSP proxy that's not propagating the logout upwards to the IdP, then there might be a bug in SSP that we should investigate.

Still, remember that proxy mode is not **officially* supported. In fact, it doesn't propagate logout messages the other way around: #65.

**Comment by tauceti2** [comment]
If SSP will be able to propagate the logout request to the upwards IdP that will be fine and most probably it will solved my problem, however this is not happening (checked via SAML tracer). Logout from IdP to the SP proxy endpoint is not need in our case.

**Comment by BAGELreflex** [comment]
I am using SSP 1.17.1. When I add the `'ForceAuthn' = true` flag to the `config/authsources.php`, I expect the metadata generated at this endpoint `/simplesaml/module.php/saml/sp/metadata.php/idp.mysite.com?output=xhtml` to contain the entry. That way when my IdP utilizes the dedicated metadata URL to refresh the metadata in my database, it knows to add the **ForceAuthn** to the metadata.

The metadata being generated does not contain this attribute. Am I mistaken that this attribute should be there? The docs state this can be both a runtime config (Parameters) and an authsources config (Options):

- https://simplesamlphp.org/docs/stable/saml:sp#section_2
- https://simplesamlphp.org/docs/stable/saml:sp#section_4

Here is my dedicated URL:
https://sp.mysite.com/simplesaml/module.php/saml/sp/metadata.php/idp.mysite.com

**Comment by tvdijen** [comment]
You are mistaken.. It will end up in the AuthnRequest, not in metadata.
